Refactor BrokerPool

adamretter opened the pull request:
Splits out the notion of several BrokerPool instances (the static part) from an instance of BrokerPool managing a pool of DBBrokers (the class instance part).

Also incorporates the changes from https://github.com/eXist-db/exist/pull/1081.
Closes https://github.com/eXist-db/exist/issues/1075
Closes https://github.com/eXist-db/exist/issues/1095

shabanovd commented:
@adamretter why you mixing things? (61d80ad8fbaa6fedd2445169a447932121996c41). That complicates review process.

shabanovd commented on src/org/exist/storage/BrokerPool.java:
```diff
@@ -1004,8 +758,9 @@ protected void initialize() throws EXistException, DatabaseConfigurationExceptio
+                            //TODO(AR) remove this!
                             //require to allow access by BrokerPool.getInstance();
-                            instances.put(instanceName, this);
+                            //instances.put(instanceName, this);
```
without that it's not possible to get `BrokerPool` by `BrokerPool.getInstance()`

adamretter replied:
@shabanovd Really... those 12 lines are complicating things for you???

adamretter replied:
Indeed. But this was a bad design decision before as it allowed someone to write code which was accesing a BrokerPool which had not finished being initialized.

Adding it to `instances` at this point is unnecessary, it will be added by `BrokerPools#configure` when initialization is complete. Separating the concerns of BrokerPool and BrokerPools as I have done enabled me to spot such a bug which otherwise was not easy to understand.

shabanovd replied:
initialization have two phases: BrokerPool startup & subsystems startup. I'm agree with you that on BrokerPool startup phase it must not go to `instances`, but disagree that on subsystems startup phase sybsystem must not have access to `BrokerPool`. At time of that phase db is up and running.

If follow your logic linux must not startup if any subsystem fail, but it's not happening.

adamretter replied:
I think how the subsystem startup in eXist done is a train-wreck, in light of what you high-lighted in `Configurator` I will give this some further thought and likely add some cleanup for that too.

adamretter commented:
@shabanovd @shabanovd Really... those 12 lines in 61d80ad are complicating things for you???
I opened this PR because you found the last one too complex!

shabanovd commented on src/org/exist/storage/BrokerPools.java:
```diff
@@ -0,0 +1,319 @@
+    // register a shutdown hook
+    static {
+        try {
+            Runtime.getRuntime().addShutdownHook(new Thread("BrokerPools-ShutdownHook") {
```
@adamretter are you sure that there is no case when it must be disabled?

adamretter replied:
@shabanovd Currently... yes I am sure of that

shabanovd replied:
@adamretter if so, why you didn't cleanup `exist.register-shutdown-hook` staff?

adamretter replied:
@shabanovd That is still used to control the Jetty shutdown hook (`JettyStart`), however that could most likely also be cleaned up to just unconditionally add the shutdown hook.

adamretter replied:
@shabanovd I have now added a commit which does that.

shabanovd commented on src/org/exist/storage/BrokerPools.java:
```diff
@@ -0,0 +1,319 @@
+                //Create the instance
+                final BrokerPool instance = new BrokerPool(instanceName, minBrokers, maxBrokers, config, statusObserver);
+                //Add it to the list
+                instances.put(instanceName, instance);
```
better to split logic here:
```
final BrokerPool instance = new BrokerPool(instanceName, minBrokers, maxBrokers, config, statusObserver);
instances.put(instanceName, instance);
instance.initialize();
```
because some subsystems require access to `BrokerPool` by `BrokerPool.getInstance()`

adamretter replied:
@shabanovd Nope that is a very bad idea, as soon as you add the BrokerPool instance to `instances` it is generally available to all threads, so you could have other threads trying to access a partially initialized BrokerPool which will lead to undefined behaviour.

When you say:
> because some subsystems require access to BrokerPool by BrokerPool.getInstance()

...my testing revealed that no subsystems prior to the completion of initialization needed to call `BrokerPool.getInstance()`. So this change is okay :-)

shabanovd replied:
@adamretter good that you are so sure, just ignore this [line](https://github.com/eXist-db/exist/blob/b476eb50b4ef32873f33190f2b4c3aa42b7f2188/src/org/exist/config/Configurator.java#L679) then

adamretter replied:
@shabanovd I didn't intentionally ignore that line obviously! The entire test suite passes. So either this is a non-issue, or it is not tested. I suspect the real culprit is the code in `Configurator` which itself looks suspect i.e.

```java
 try {
    db = BrokerPool.getInstance();
} catch (final EXistException e) {
   //ignore if database is starting-up
   //TODO: add to BrokerPool static list to activate when ready
}
```

adamretter replied:
@shabanovd I just did some testing and in reality that code in `Configurator` does not appear to be called until after the `BrokerPool` is initialized and added to the `instances`, if at all

shabanovd replied:
@adamretter that happen because you have anything that use this part of code. You should add realm for example.

adamretter replied:
@shabanovd Okay I see the issue for `Realm`, although it doesn't stop the Realm from starting (as it will subsequently be started by the SecurityMangerImpl), so in reality all works well.

I would ask, in general, should the `Configurator` really be starting sub-modules???

Also it seems this is a mess as a `Realm` would before this PR have been started twice (unnecessarily):

1. Line 688 of `Configurator`:
    ```java
    if (db != null) {
        try(final DBBroker broker = db.getBroker()) {
            ((LifeCycle) obj).start(broker);
        }
    }
```

2. Line 200 of `SecurityManagerImpl`:
    ```java
        for (final Realm realm : realms) {
            realm.start(broker);
        }
```

shabanovd replied:
Line 200 is workaround for issue I describe above. I add proposal to #1085 to resolve it. And I don't think "passing them the BrokerPool" will work in all cases.

shabanovd replied:
@adamretter are you telling "Plug which seems to be unused" just because you do not use it?

adamretter replied:
@shabanovd No I am saying that I didn't see anything in the eXist codebase that uses it.

shabanovd replied:
@adamretter eXist codebase don't have all code of eXist ecosystem

adamretter replied:
I am aware of that, but I don't know of any projects from "eXist-db" which use it either...

shabanovd replied:
also eXist codebase have `MDStorageManager` that use `Plug` but you simply ignore that extention (and creating your own bike)

adamretter replied:
@shabanovd `MDStorageManager#start` does nothing, so it is not affected by this change.

shabanovd replied:
@adamretter Was I be clear enough? I do not like your "Plug which seems to be unused" and putting counterargument!!!

adamretter replied:
@shabanovd That is fine.

wolfgangmm replied:
The problem still is that startup triggers will fail with this PR and we need to solve this somehow. The exceptions are clear: access to the security manager requires BrokerPool.getInstance to return a valid instance. So it's not possible to create users during startup.

I'm not sure how to solve it. Maybe startup triggers should be run from BrokerPools once it has configured a pool instead? Will try this...

```
org.expath.pkg.repo.PackageException: Failed to create user: eXide
        at org.exist.repo.Deployment.checkUserSettings(Deployment.java:620) ~[exist.jar:?]
        at org.exist.repo.Deployment.deploy(Deployment.java:405) ~[exist.jar:?]
        at org.exist.repo.Deployment.installAndDeploy(Deployment.java:271) ~[exist.jar:?]
        at org.exist.repo.AutoDeploymentTrigger.execute(AutoDeploymentTrigger.java:72) [exist.jar:?]
        at org.exist.storage.BrokerPool.callStartupTriggers(BrokerPool.java:774) [exist.jar:?]
        at org.exist.storage.BrokerPool.initialize(BrokerPool.java:686) [exist.jar:?]
        at org.exist.storage.BrokerPool.<init>(BrokerPool.java:398) [exist.jar:?]
        at org.exist.storage.BrokerPools.configure(BrokerPools.java:178) [exist.jar:?]
        at org.exist.storage.BrokerPools.configure(BrokerPools.java:110) [exist.jar:?]
        at org.exist.jetty.JettyStart.run(JettyStart.java:163) [exist-optional.jar:?]
        at org.exist.launcher.Launcher$2.run(Launcher.java:176) [exist.jar:?]
Caused by: org.exist.config.ConfigurationException: org.exist.EXistException: Database instance 'exist' is not available
        at org.exist.security.AbstractPrincipal.<init>(AbstractPrincipal.java:68) ~[exist.jar:?]
        at org.exist.security.AbstractGroup.<init>(AbstractGroup.java:56) ~[exist.jar:?]
        at org.exist.security.internal.GroupImpl.<init>(GroupImpl.java:49) ~[exist.jar:?]
        at org.exist.security.internal.SecurityManagerImpl.addGroup(SecurityManagerImpl.java:561) ~[exist.jar:?]
        at org.exist.repo.Deployment.checkUserSettings(Deployment.java:608) ~[exist.jar:?]
```

adamretter replied:
The solution is to change the startup of subsystems. We need a `prepare` method that happens after `configure` but before `initialize`. It should be triggered from the tail of `configure` and should only take a single argument which is the `BrokerPool`. That `BrokerPool` should then be explicitly passed to each module which needs to be prepared before the BrokerPool instance is made generally available.

I did open an issue for this https://github.com/eXist-db/exist/issues/1085 here, but I did not work on it yet as I didn't see any problems yet with Startup Triggers. I guess there are no tests around auto Package Deployment?

If you could send a PR with a test for the issue you are seeing, then I could likely amend this PR with a commit this evening to fix the startup issue you have with StartupTriggers.

shabanovd commented on src/org/exist/storage/BrokerPools.java:
```diff
@@ -0,0 +1,319 @@
+                //TODO : call isConfigured(id) and throw an EXistException if relevant ?
+                return instance;
+            } else {
+                throw new EXistException("Database instance '" + instanceName + "' is not available");
```
@adamretter there was `instancesInitializtionException` that unhide initialization exception, why you did remove it?

adamretter replied:
Okay so when I refactored, that approach revealed two bugs / design issues:

1) Previously, if you attempted to configure an instance and it failed, you could never reattempt this again for the same *instance name*, even if you provided a completely different/fixed config. This prevented the ability to retry or recover from a bad configuration at runtime.

2) Keeping the initalization exceptions was only necessary to enforce (1). The function signature of configure reveals that it can throw an `EXistException` and previously it never did, instead that would be thrown at `getInstance` time, there were TODO comments suggesting that this should be corrected; Correcting the issue has simplified the code, removed deferred error-reporting of initialization exceptions and reduced memory, as well as fixing (1).

shabanovd replied:
I don't know where you find case 1. `instancesInitializtionException` get into game only when `instances` have nothing for given name.

With current code you hide actual exception from end-user by "is not available" exception (that gives nothing and force to ask user to check logs)

adamretter replied:
That's not the case. Now `is not available` means exactly that, i.e. you have not configured the instance for the name you provided. You will have received the exception in the `configure` call, so you don't have to check the logs. Everything works and the test suite passes, so I don't think this is a *real* issue.

shabanovd replied:
Hardcode any exception at initialization block and run java client and run it by script. What do you see?

adamretter replied:
@shabanovd I see exactly the same error as if it had been thrown from `getInstance` instead of `configure`. What is the problem you are alluding to?

shabanovd replied:
@adamretter is it "is not available" exception or actual exception that happen at initialization time?

adamretter replied:
@shabanovd it's the actual exception that happened at initialization time.

shabanovd replied:
@adamretter good if so, and mails like [this one](https://sourceforge.net/p/exist/mailman/message/26993753/) (with huge none informatice exception) will not comeback.

adamretter replied:
@That email is from 5 years ago! There have been many changes to eXist (and BrokerPool) since then. If there is a genuine issue, I will look into it.
Is there a genuine issue that you have?

shabanovd replied:
@adamretter I did add `instancesInitializtionException` by 7020234bb13dfd6ea933f9d99a1bebbb9f4e205c to get more informative exception. Now, you are saying that it's not a case anymore because it was 5 years ago, why?

shabanovd commented on extensions/debuggee/src/org/exist/debuggee/ScriptRunner.java:
```diff
@@ -69,7 +69,7 @@ public void run() {
 			final Database db = BrokerPool.getInstance();
 			
-            BrokerPool.registerStatusObserver(this);
+//            BrokerPool.registerStatusObserver(this);
```
@adamretter without this debuggee session will not be closed

adamretter replied:
AFAIK the Debuggee code does not work and is unsupported, which is why I did not spend too much time on that, am I mistaken?

If that is not the case, then each instance of BrokerPool now has a `status observer`, we could use that here if you like?

shabanovd replied:
@adamretter good to know that Debuggee was thrown out, when it happen?

adamretter replied:
@shabanovd I didn't say it was "thrown out", I said it was unsupported. AFAIK you added this code and since then no one has used it because the eXist-db website says "This is functional, but incomplete (the client side in particular)".

shabanovd replied:
@adamretter "incomplete" and "unsupported" totally different things, no?

adamretter replied:
@shabanovd yes. Are you supporting it?

shabanovd replied:
@adamretter Is it not seen from my defense of this code? I did code it and I didn't say that do not support it anymore.

adamretter replied:
@shabanovd Okay that is good news then :-)

shabanovd commented:
@adamretter that is not just "those 12 lines", but you change many things instead of extract instances and add lock.

dizzzz commented:
I need some time to understand the message thread and the changes ....

adamretter commented:
@shabanovd I have rebased and fixed the Debugee Script Runner as requested so that the observer is respected. I have also added a commit to cleanup the shutdown hook stuff as you requested. I now believe that I have addressed all your concerns that I see as valid.

@dizzzz I am happy with this now.

shabanovd commented on extensions/debuggee/src/org/exist/debuggee/ScriptRunner.java:
```diff
@@ -68,8 +68,8 @@ public void stop() {
-			
-            BrokerPool.registerStatusObserver(this);
+
+            db.addStatusObserver(this);
```
should it be unregistered too?

adamretter replied:
@shabanovd You never unregistered it in the past... so I guess that is something for you to answer!?

shabanovd replied:
@adamretter I ask you because you did redesign, I should do that?

adamretter replied:
@shabanovd nothing has really changed around observers... if ***you*** didn't need to unregister **your** observer in the past, then you don't need to now either

shabanovd replied:
@adamretter will it lead to memory leak?

adamretter replied:
@shabanovd No

shabanovd replied:
@adamretter ok

dizzzz commented:
interesting PR,, good it solves this issue I have found, interesting... that you unhide design 'defects' all the time :-) 

@ljo @wolfgangmm  please could you check? IMO this should be in 3.0 final...

shabanovd commented:
lgtm

dizzzz commented:
Thnx @shabanovd ! @adamretter the change cannot be merged in....

adamretter commented:
@dizzzz Why do you say "the change cannot be merged in..." there are no conflicts!

shabanovd commented:
@adamretter because "Merge pull request" was disabled yesterday

adamretter commented:
@shabanovd er... How do you mean disabled? Was it a problem with GItHub?

shabanovd commented:
@adamretter hard to say, I saw it disabled for several PR here

adamretter commented:
@shabanovd Do you still see it disabled? It is visible for me...

shabanovd commented:
@adamretter yes, it's ok now

dizzzz commented:
Yep it is ok again

adamretter commented:
@shabanovd @dizzzz okay great... How about pushing that button then? ;-)

wolfgangmm commented:
I'm retesting on windows where I had some other shutdown/restart issues before. Will merge if it works.

wolfgangmm commented:
I found minor issues with this PR which require some small changes. Working on it...

shabanovd commented:
@wolfgangmm any details? (shortly, if possible)

wolfgangmm commented:
* I'm getting an exception during the test suite in org.exist.start.Main.shutdown:

[junit] java.lang.IllegalAccessException: Class org.exist.start.Main can not access a member of class org.exist.storage.BrokerPools with modifiers "public static"

* Restarting eXist via the launcher failed in some cases (already fixed)

adamretter commented:
If this is the test-suite. I am wondering why I don't see the problem when I run the test-suite here?

wolfgangmm commented:
#1095 extends this PR with a test case, a fix for the startup triggers and a shutdown issue.

adamretter commented:
@shabanovd I knew you would be very excited about these changes. However, please don't review just yet as this is still a work in progress; I needed a backup. I will let you knowehen I drop it...

shabanovd commented:
@adamretter ok

dizzzz commented:
what is the status of the PR?

adamretter commented:
@dizzzz Status is almost done... but not quite yet. Should complete it tomorrow...

dizzzz commented:
Adam? any update?

adamretter commented:
@dizzzz Hoping to complete today/tomorrow. I had to fix some issues with serialization and kwic for a client first.

dizzzz commented:
@adamretter thnx for the headsup!

adamretter commented:
@dizzzz @wolfgangmm Okay it is ready to go, please test and merge. It includes the necessary fixes that @wolfgangmm made in #1082.

shabanovd commented on .classpath:
```diff
@@ -25,6 +25,7 @@
 	<classpathentry kind="src" path="extensions/debuggee/test"/>
 	<classpathentry kind="src" path="extensions/webdav/src"/>
 	<classpathentry kind="src" path="extensions/commands/src"/>
+	<classpathentry kind="src" path="lib/core/j8fu-1.0.jar"/>
```
what is license it have?

adamretter replied:
BSD-3 Clause - https://repo1.maven.org/maven2/com/evolvedbinary/j8fu/j8fu/1.0/j8fu-1.0.pom

shabanovd commented on extensions/security/oauth/src/org/exist/security/realm/oauth/OAuthRealm.java:
```diff
@@ -44,7 +44,7 @@
 import org.exist.security.internal.aider.UserAider;
 import org.exist.storage.DBBroker;
-import org.exist.util.function.Function2E;
+import com.evolvedbinary.j8fu.function.Function2E;
```
why you replace eXist class by external?

adamretter replied:
Actually it's kinda the other way. I had always copied and pasted simpler versions of some functional utilities that I had into eXist, it got to the point where I was maintaining them in two places and the upstream (j8fu) is more complete, so I just decided to add the dependency instead of copy/pasting and doing things twice.

shabanovd commented on src/org/exist/collections/CollectionCache.java:
```diff
@@ -38,13 +39,13 @@
 
     public CollectionCache(BrokerPool pool, int blockBuffers, double growthThreshold) {
-        super(blockBuffers, 2.0, 0.000001, CacheManager.DATA_CACHE);
+        super(blockBuffers, 2.0, growthThreshold, CacheManager.DATA_CACHE);
```
are you changing value from `0.000001` to `0.0001`?

adamretter replied:
Not exactly previously the `growthThreshold` parameter was orphaned, and it was clear that that should not have been the case.

adamretter replied:
@shabanovd Yes that is correct, however from how the code was written, I suspect the growthThreshold was meant to be `0.0001`: https://github.com/eXist-db/exist/blob/develop/src/org/exist/storage/BrokerPool.java#L840

Do you agree?

shabanovd replied:
my question is simple: how well you stress test db with that new parameter?

If you did not stress test db then change line 840 at `BrokerPool` to keep old parameter.

adamretter replied:
@shabanovd There is no common stress-test facility for eXist. I could run some stress-tests but they would only be representative of a specific artificial workload, which almost certainly would not be the same as your workload.

I will change this to the previous value of `0.000001`...

adamretter replied:
@shabanovd Done, rebased and pushed. It's ready?!?

adamretter commented on src/org/exist/plugin/PluginsManagerImpl.java:
```diff
@@ -78,21 +83,31 @@
+	@Override
 	public void start(DBBroker broker) throws EXistException {
-        final TransactionManager transaction = db.getTransactionManager();
+        final TransactionManager transaction = broker.getBrokerPool().getTransactionManager();
```
@shabanovd That is the name you gave it when you created this code, I didn't change the name. Would you like me to fix it?

dizzzz commented:
Wooohooo

dizzzz commented:
for me.... the tests do not finish it seems :-/  last test is org.exist.indexing.ngram.MatchListenerTest

yourkit does not detect a deadlock though

adamretter commented:
Hmm they finish on my Mac and on Travis can you send the stack output
please?

On Friday, 7 October 2016, Dannes Wessels <notifications@github.com> wrote:

> for me.... the tests do not finish it seems :-/ last test is
> org.exist.indexing.ngram.MatchListenerTest
>
> yourkit does not detect a deadlock though
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/eXist-db/exist/pull/1082#issuecomment-252345095>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABNJuVZ3a33aLMxy7iFIGpUd5AOTVp0rks5qxqJQgaJpZM4JdjWe>
> .
>


-- 
Adam Retter

skype: adam.retter
tweet: adamretter
http://www.adamretter.org.uk

dizzzz commented:
jps
1874 Jps
1739 JUnitTestRunner
1517 Launcher


[launcher.txt](https://github.com/eXist-db/exist/files/517015/launcher.txt)
[testrunner.txt](https://github.com/eXist-db/exist/files/517016/testrunner.txt)


git status
On branch broker-pool-rewrite
Your branch is up-to-date with 'origin/broker-pool-rewrite'

git rev-parse HEAD
2a5e4e5e8cf9c417d01119c3c7cfbd85bd6bd67a

adamretter commented:
@dizzzz I think during shutdown a SystemTask job which was scheduled execute (perhaps Sync? I can't tell without looking at your exist.log) and that causes a deadlock between the SecurityManager and the Database shutdown. Just a theory at the moment, I will dive in more tomorrow...

dizzzz commented:
attached exist.log

[exist.log.gz](https://github.com/eXist-db/exist/files/517113/exist.log.gz)

adamretter commented:
@dizzzz I just pushed some fixes around shutdown and System Tasks. Can you try and run it again on your system.

If it still deadlocks, can you send me an updated `jstack` output for the `JUnitTestRunner` process and also your new `exist.log`?
